scripts/o3de/o3de/compatibility.py

```diff
@@ -106,127 +87,104 @@ def gem_project_compatible(gem_json_data:dict, project_path:pathlib.Path, gem_pa
-            engine_json_data)
-    return True
-
-def engine_is_compatible(engine_name:str, engine_version:str, compatible_engines:list, engine_api_version_specifiers:list, engine_json_data:dict or None) -> bool:
-    """
-    Returns True if the engine is compatible with the provided compatible_engines and engine_api_versions information.
+    Returns any incompatible objects for this object and engine.
     If a compatible_engine entry only has an engine name, it is assumed compatible with every engine version with that name.
     If an engine_api_version entry only has an api name, it is assumed compatible with every engine api version with that name.
-    :param engine_name: the engine name
-    :param engine_version: the engine version
-    :param compatible_engines: a list of engine names and (optional)version specifiers
-    :param engine_api_versions: a list of engine api names and (optional)version specifiers
-    :param engine_json_data: optional engine json data to avoid extra file ops
+    :param object_json_data: the json data for the object 
+    :param engine_json_data: the json data for the engine
     """
+    compatible_engines = object_json_data.get('compatible_engines','')
+    engine_api_version_specifiers = object_json_data.get('engine_api_dependencies','')
+
     # early out if there are no restrictions
     if not compatible_engines and not engine_api_version_specifiers:
-        return True
-
-    if not engine_json_data:
-        if engine_name:
-            engine_json_data = manifest.get_engine_json_data(engine_name=engine_name)
-        else:
-            engine_json_data = manifest.get_engine_json_data(engine_path=manifest.get_this_engine_path())
-
-    if not engine_json_data:
-        return False
+        return set() 
 
-    if not engine_name:
-        # fallback to use the information from the current engine
-        engine_name = engine_json_data.get('engine_name','')
-        engine_version = engine_json_data.get('version','')
+    engine_name = engine_json_data['engine_name']
+    engine_version = engine_json_data.get('version','')
 
+    incompatible_objects = set()
     if compatible_engines:
         if engine_version and has_compatible_version(compatible_engines, engine_name, engine_version):
-            return True
+            return set() 
         elif not engine_version and has_compatible_name(compatible_engines, engine_name):
             # assume an engine with no version is compatible
-            return True
+            return set()
 
-    engine_api_versions = engine_json_data.get('api_versions','')
-    if engine_api_version_specifiers and not engine_api_versions:
-        # assume not compatible if no engine api version information is available
-        return False
+        # object is not known compatible with this engine
+        incompatible_objects.add(f"{engine_name} {engine_version} does not match any version specifiers in the list of compatible engines: {compatible_engines}")
```

Comment:
Is this meant to be a string or the actual incompatible object?

Reply:
each incompatible object is a human readable string that should help the user determine why something isn't compatible.  For the engine, I chose to add a bit more help text to give the user as much information as possible.

Reply:
So it is pretty much the error part of a structure like an `AZ::Outcome<void, AZStd::string>`